lib/elixir/lib/list.ex

```diff
@@ -577,22 +577,51 @@ defmodule List do
   end
 
   @doc """
-  Returns `true` if `list` starts with the given `prefix` list.
+  Returns `true` if `list` starts with the given `prefix` list; otherwise returns `false`.
+
+  `list` and `prefix` can be proper or improper lists.
+
+  If `prefix` is an empty list, it returns `true`.
 
   ### Examples
 
       iex> List.starts_with?([1, 2, 3], [1, 2])
       true
+
+      iex> List.starts_with?([1, 2], [1, 2, 3])
+      false
+
       iex> List.starts_with?([:alpha], [])
       true
+
       iex> List.starts_with?([], [:alpha])
       false
 
+      iex> List.starts_with?([1, 2 | 3], [1, 2])
+      true
+
+      iex> List.starts_with?([1, 2 | 3], [1, 2 | 3])
+      true
+
+      iex> List.starts_with?([1, 2, 3], [1, 2 | 3])
+      false
+
   """
-  @spec starts_with?(list, list) :: boolean
-  def starts_with?(list, prefix) do
-    :lists.prefix(prefix, list)
-  end
+  @spec starts_with?(maybe_improper_list, maybe_improper_list) :: boolean
+  @spec starts_with?(maybe_improper_list, []) :: true
+  @spec starts_with?([], nonempty_maybe_improper_list) :: false
+  def starts_with?(list, prefix)
+  def starts_with?(list, list) when is_list(list),
```

Comment:
@michalmuskala thank you for suggesting that. I will take it into account for future times.

@josevalim and @whatyouhide The whole idea of re-implementing it was because our needs were a bit different than the ones covered by the Erlang version. We agreed on the mailing list to cover improper lists in the first argument, and do not worry about improper lists as the second argument to not to add complexity if it's not required. The Erlang version doesn't deal with improper lists at all. The thing is that while implementing this code I realized that it is really difficult  to add support for only the first arg. and not the second one, but it is pretty straight forward supporting both args.

Also as Andrea's mentioned, the error message is not very descriptive and it refers to an Erlang function with a complete different name.

I'm merging a few changes that take @michalmuskala's comment into account and it should deal properly with both proper and improper lists. have a look and tell me what you think.